tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -131,29 +130,33 @@ def get_config(self):
         return config
 
     @classmethod
-    def create_optimizer_spec(cls, optimizer_instance, layer):
-
-        assert isinstance(
-            optimizer_instance, tf.keras.optimizers.Optimizer
-        ), "Object passed is not an instance of tf.keras.optimizers.Optimizer"
-
-        assert isinstance(layer, tf.keras.layers.Layer) or isinstance(
-            layer, tf.keras.Model
-        ), "Object passed is not an instance of tf.keras.layers.Layer nor tf.keras.Model"
+    def create_optimizer_spec(
+        cls,
+        optimizer: tf.keras.optimizers.Optimizer,
+        layer: Union[
```

Comment:
Also the naming here It Is a little bit confusing also in the internal code cause if `layers` with `s` It could cover the `list` case but what about `tf.keras.Model`?

Reply:
Sure, the naming is not so good to me either, but I cannot come up with one... do you have any suggestion?

Reply:
Ugly as ugly we already have `optimizers_and_layers`. 
With another boolean we could have `layers_or_model` :smile:

Reply:
Can you share the input signature you propose?

Reply:
that's a good idea. please go ahead with the renaming.

Reply:
`layer` -> `layers_or_model`

Reply:
Hi, seems that `tf.keras.Model` is a subclass of `tf.keras.layers.Layer`. Do we still need to do this?

Reply:
It is a borderline case cause is multi inheritance:
```class Model(base_layer.Layer, version_utils.ModelVersionSelector):```
We are only using `layers` so a single base classe features but I don't know about readibility.

Reply:
So as you want cause we cannot upcast to `Layer` in python or It Is generally a [borderline pratice](https://docs.python.org/3.8/reference/datamodel.html#id1)

Reply:
Updated the naming. See if this is better now :-)